### ISSUE TITLE
ProbeNode: rotation results in incorrect default pointer areas

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/scenery-phet#488]
Documented in https://github.com/phetsims/beers-law-lab/issues/236.  Also seen in the Concentration sim.  When looking at the pointer areas of the concentration tool, the blue one that indicates mouse interaction droops down.  On my iphone it seemed to occur in landscape mode, but upon checking other platforms it seems to be the norm there.  This pointer area doesn't seem to be actually be interactable and the tool itself can be moved normally.  As such, there is no pedagogical impact and this is likely a minor issue.

Found on the published versions of the sims.

**Comment by pixelzoom** [comment]
I don't see anything obviously wrong with ProbeNode.  @jonathanolson is a Node's transform applied to its pointer areas?

**Comment by jonathanolson** [comment]
> I don't see anything obviously wrong with ProbeNode. @jonathanolson is a Node's transform applied to its pointer areas?

The touch/mouse area are in the local coordinate frame, so generally no

**Comment by samreid** [comment]
I tested https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482180713 using this patch:

```
Index: js/demo/ComponentsScreenView.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/demo/ComponentsScreenView.js	(revision c91007a5bbebb999ce4c0be0a3bf366a52d29571)
+++ js/demo/ComponentsScreenView.js	(date 1555003558000)
@@ -12,9 +12,10 @@
   'use strict';
 
   // modules
-  var BooleanProperty = require( 'AXON/BooleanProperty' );
+  var SimpleDragHandler = require( 'SCENERY/input/SimpleDragHandler' );
   var ArrowKeyNode = require( 'SCENERY_PHET/keyboard/ArrowKeyNode' );
   var ArrowNode = require( 'SCENERY_PHET/ArrowNode' );
+  var BooleanProperty = require( 'AXON/BooleanProperty' );
   var BracketNode = require( 'SCENERY_PHET/BracketNode' );
   var CapsLockKeyNode = require( 'SCENERY_PHET/keyboard/CapsLockKeyNode' );
   var Checkbox = require( 'SUN/Checkbox' );
@@ -39,10 +40,10 @@
   var HandleNode = require( 'SCENERY_PHET/HandleNode' );
   var HBox = require( 'SCENERY/nodes/HBox' );
   var HeaterCoolerNode = require( 'SCENERY_PHET/HeaterCoolerNode' );
-  var KeyboardHelpSection = require( 'SCENERY_PHET/keyboard/help/KeyboardHelpSection' );
   var HSlider = require( 'SUN/HSlider' );
   var HStrut = require( 'SCENERY/nodes/HStrut' );
   var inherit = require( 'PHET_CORE/inherit' );
+  var KeyboardHelpSection = require( 'SCENERY_PHET/keyboard/help/KeyboardHelpSection' );
   var Keypad = require( 'SCENERY_PHET/keypad/Keypad' );
   var KitSelectionNode = require( 'SCENERY_PHET/KitSelectionNode' );
   var LaserPointerNode = require( 'SCENERY_PHET/LaserPointerNode' );
@@ -529,7 +530,7 @@
       ],
       function() {
         probeNodeLayer.removeAllChildren();
-        probeNodeLayer.addChild( new ProbeNode( {
+        const probeNode = new ProbeNode( {
 
           // ProbeNode options
           color: colorProperty.value,
@@ -543,8 +544,14 @@
 
           // layout options
           x: layoutBounds.centerX,
-          y: layoutBounds.centerY
+          y: layoutBounds.centerY,
+          rotation: -Math.PI / 2,
+          cursor: 'pointer'
+        } );
+        probeNode.addInputListener( new SimpleDragHandler( {
+          start: function() {}
         } ) );
+        probeNodeLayer.addChild( probeNode );
       } );
 
     // Show a cross hairs in the middle of the screen so that we can verify that the sensor's origin is correct.

```

And found that the hand turned into a cursor over the probe itself, not over the purple/blue area.

**Comment by pixelzoom** [comment]
I'm going to take a stab at debugging this in scenery playground. Some notes to myself:

* http://phetsims.github.io/scenery/tests/playground.html
* `scene` is the top-level node
* namespacing is different, e.g. `window.dot`, `window.kite`, `window.scenery`, etc.
* to show pointer areas, `display.setPointerAreaDisplayVisible( true )`
* `display.updateDisplay()` is not automatically called; call it explicitly, or press the "Update" button 

Example:

```
> display.setPointerAreaDisplayVisible( true )
> var rect = new scenery.Rectangle( 0, 0, 100, 100, { fill: 'red', left: 100, top: 100 } )
> scene.addChild( rect )
> rect.touchArea = rect.localBounds.dilatedXY( 10, 10 )
> rect.rotation = Math.PI/4
```

**Comment by pixelzoom** [comment]
Here's where the pointer areas are set in ProbeNode:

```js
    var outline = createShape().close();
    options.mouseArea = options.mouseArea || outline;
    options.touchArea = options.touchArea || outline;
```

If I add 2 calls to `outline.copy()`, the problem goes away:

```js
    var outline = createShape().close();
    options.mouseArea = options.mouseArea || outline.copy();
    options.touchArea = options.touchArea || outline.copy();
```

So I thought that perhaps each pointer area needed a unique shape.  But I tried this, and it does not work:

```js
    options.mouseArea = options.mouseArea || createShape().close();
    options.touchArea = options.touchArea || createShape().close();
```

So I suspect that there's something funky about the Shape created by `createShape().close()`, and that the funkiness is resolved by calling `copy`.   

@jonathanolson could you take a peek?

**Comment by jonathanolson** [comment]
I'll take a look.

**Comment by pixelzoom** [comment]
@jonathanolson In case it's significant, note @samreid's findings in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482215760.  For mouseArea, the cursor changes over the expected area.  So it may just be that the pointer area shapes are drawn incorrectly. 

Also note that part of the shape is drawn correctly, and part (the partial ellipse?) is not.

**Comment by KatieWoe** [comment]
I saw that pH Scale Basics has an odd  red pointer area that starts just off screen, connected to the Probe Node, that can be manipulated when on screen.  Seems like it may be related.
![Image from iOS](https://user-images.githubusercontent.com/41024075/56069267-83f97f80-5d3f-11e9-9167-10d693e9dc90.png)

**Comment by pixelzoom** [comment]
@ariel-phet I had planned to create an RC (new release branch) for _pH-Scale_ and _pH Scale: Basics_ this coming Monday.  But since this problem is occurring in those sims, I'm going to hold off.  Please prioritize this issue, and let me know how you want me to proceed with RC.

**Comment by pixelzoom** [comment]
A couple of things to note about the _pH Scale_ case that @KatieWoe identified in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482741494:
* pH Scale rotation is `Math.PI/2`, while the BLL rotation was `-Math.PI/2`
* pH Scale shows a problem with `touchArea` specified by the client, while BLL shows a problem with ProbeNode's default `touchArea`

**Comment by pixelzoom** [comment]
A couple more clues:

* in BLL, the problem is transient; it isn't visible until you move the probe around, and it comes and goes. In pH Scale, the problem is persistent.

* The problem is present in the published versions of both sims.

**Comment by pixelzoom** [comment]
This problem seems to be present in any sim that uses a rotated ProbeNode.   In addition to beers-law-lab and ph-scale, that includes energy-skate-park (@samreid @jessegreenberg), see screenshot below.

@KatieWoe if you find other sims that are affected, please note in this issue, rather than sim-specific issues.

<img width="1038" alt="screenshot_1135" src="https://user-images.githubusercontent.com/3046552/56081299-ff534380-5dc8-11e9-9665-527f5b1f0a41.png">

**Comment by ariel-phet** [comment]
Although this issue is odd, (and is worthy of investigation), it does not appear to be causing any noticeable issues (and it has existed for a long time without complaint in the published sims). It would be good to fix in the long run, but I do not see it as a blocking issue. 

@pixelzoom you are free to proceed with RC's as you see fit.

**Comment by KatieWoe** [comment]
I would generally agree.  I'm a bit more concerned about this aspect: https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482741494 if it is connected.  If it's not, let me know and I will make a new issue.

**Comment by ariel-phet** [comment]
@KatieWoe that aspect is definitely a bit funky, but I think still not a blocking issue as it seems unlikely to cause any strange use cases. I can actually grab in this spot using touch, but only when the probe is in a place where it would be pretty unlikely to be touching the screen where the "cord" touch area appears. Would definitely be worth understanding and fixing, but still not blocking or high priority.

**Comment by pixelzoom** [comment]
If it was simply the case that pointer areas are drawn in the wrong place, I would be more comfortable with proceeding.  But the fact that you can grab the incorrect pointer area is disconcerting.  And my workaround for BLL unfortunately does not apply in general, i.e. does not fix pH Scale.

In Slack discussion with @ariel-phet, he indicated that @jonathanolson's status on this issue is:

> He said he will be looking into it today
> 
> But would also be happy to push the fix to your release branch if there is one

I will wait until tomorrow 4/16 before proceeding with release branches for pH Scale and pH Scale: Basics.

**Comment by jonathanolson** [comment]
I'm curious how I can "grab" in the incorrect area, or reproduce this? So far it seems like it's mainly a graphical glitch in how things are being shown with the query parameter.

**Comment by KatieWoe** [comment]
I don't believe you can grab the node, or at least I have not seen it.  But the red box in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482741494 can be grabbed and moved.

**Comment by samreid** [comment]
While I was trying to reproduce the problem with the red box in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482741494 I noticed that the pointer area for the ProbeNode flickers during window resize:

![flicker](https://user-images.githubusercontent.com/679486/56226820-4d6a8000-6031-11e9-956b-1e0cd892ed5b.gif)

Kind of hard to tell from the gif, but it flickers between the correct area and the "rotated up" area.

**Comment by samreid** [comment]
I confirmed I can also drag the probe by its red rectangle in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-482741494.   I also noted that when dragging the red rectangle, there are certain regions of the screen where the ProbeNode pointer area goes buggy, and some regions where it is normal.

**Comment by jonathanolson** [comment]
I'm mostly confirming that this is manifesting only as a display-only problem.

It looks like `Shape.transformed` is not behaving correctly for a certain class of transforms (depending on the exact scale/translation) for the probe shape. This is used for the display, but NOT in the hit testing.

For instance a buggy case is the Shape:
```
M 0.00000000000000000000 59.00000000000000000000 L -2.99999999999999911182 59.00000000000000000000 A 12.00000000000000000000 12.00000000000000000000 0 0 1 -15.00000000000000000000 47.00000000000000000000 L -15.00000000000000000000 44.00000000000000000000 Q -15.00000000000000000000 34.00000000000000000000 -27.50657780874820801387 19.98469857794409065832 A 34.00000000000000000000 34.00000000000000000000 0 1 1 27.50657780874821156658 19.98469857794408355289 Q 15.00000000000000000000 34.00000000000000000000 15.00000000000000000000 44.00000000000000000000 L 15.00000000000000000000 47.00000000000000000000 A 12.00000000000000000000 12.00000000000000000000 0 0 1 3.00000000000000088818 59.00000000000000000000 L 0.00000000000000000000 59.00000000000000000000 Z
```

Transforming with the matrix:
```
4.55902603864401e-17 -0.7445454545454545 274.80218736910194
0.7445454545454545 4.55902603864401e-17 460.99406445080257
0 0 1
```

ends with the buggy case:
```
M 230.87400555092011700253 460.99406445080256844449 L 230.87400555092011700253 458.76042808716618992548 A 8.93454545454545367988 8.93454545454545367988 0 0 1 239.80855100546557423513 449.82588263262073269289 L 242.04218736910195275414 449.82588263262073269289 Q 249.48764191455649097406 449.82588263262073269289 259.92267088243266925929 440.51416697319820059420 A 25.31454545454545268512 25.31454545454545268512 0 1 1 289.68170385577121805909 440.51416697319820059420 Q 249.48764191455649097406 472.16224626898440419609 242.04218736910195275414 472.16224626898440419609 L 239.80855100546557423513 472.16224626898440419609 A 8.93454545454545367988 8.93454545454545367988 0 0 1 230.87400555092011700253 463.22770081443894696349 L 230.87400555092011700253 460.99406445080256844449 Z
```

**Comment by pixelzoom** [comment]
@jonathanolson said:
> I'm mostly confirming that this is manifesting only as a display-only problem.

It doesn't sound like that's the case. @ariel-phet (https://github.com/phetsims/scenery-phet/issues/488#issuecomment-483367487) and @samreid (https://github.com/phetsims/scenery-phet/issues/488#issuecomment-483733447) have both reported being able to interact with the incorrect pointer areas.

**Comment by jonathanolson** [comment]
I believe @samreid's comment in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-483733447 is referring to the rectangular hit area:

<img width="146" alt="Screen Shot 2019-04-16 at 11 46 45 AM" src="https://user-images.githubusercontent.com/480878/56232166-5c572f80-603d-11e9-9996-40212ea147f2.png">

@ariel-phet can you provide an example where it is possible to click on the "improper" area? I tested with touch enabled and was unable to reproduce.

**Comment by KatieWoe** [comment]
Was looking at ph scale, and for some reason the rotating pointer area wasn't appearing.  The red dragable box was.

**Comment by pixelzoom** [comment]
@KatieWoe By "rotating pointer area" do you mean the blue dashed shape? FYI, red is touch area, blue is mouse area.

**Comment by KatieWoe** [comment]
Yes, the mouse area was not in the incorrect rotated state.

**Comment by pixelzoom** [comment]
In pH Scale, I see the mouse area becoming incorrect only after it's dragged.  @KatieWoe is that what you see too, or are you seeing that it remains correct?

**Comment by KatieWoe** [comment]
I was not seeing it at all on pH Scale for some reason.  On Win 10 chrome or on iOS 12.2 safari phone.  It seemed very odd to me since it showed up in related sims.

**Comment by ariel-phet** [comment]
@jonathanolson this is fairly easy for me to reproduce. Tried on both my windows laptop touchscreen and through the app on my phone. Basically just get the sim in this configuration:

![phscale](https://user-images.githubusercontent.com/4051174/56687266-1cbec200-6693-11e9-8081-fdd92b7946da.PNG)

Then put your finger in near the "sag" in the wire in the button left, and you can move the probe around.

Again, I don't see this as a major issue. Though easy to reproduce it is not a likely use case scenario, but it would be good to figure out why this other touch area appears.

**Comment by pixelzoom** [comment]
4/25/19 dev meeting:

@ariel-phet requested that this issue be deferred and assigned to him.

**Comment by pixelzoom** [comment]
@jonathanolson @ariel-phet FYI, the bogus touchArea (red rectangle) shown in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-486389051 was a sim-specific bug.  It has been fixed, see https://github.com/phetsims/ph-scale/issues/77.

So the remaining problem here is the default pointer areas in ProbeNode. And the cause was identified by @jonathanolson in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-483738588.

**Comment by ariel-phet** [comment]
@KatieWoe since this bug just shows something odd in a internally used tool, I am inclined to close this issue as `wontfix` and just have it be a known issue.

Feel free to close if you agree.

**Comment by pixelzoom** [comment]
Reopening.  While I agree that it's OK to ignore the problem as it manifests in ProbeNode, I don't think it's a good idea to ignore the underlying problem in kite.  Kite is a foundational repo, and that problem could easily manifest in other places.  So I'd like to see an issue that identifies the problem in kite, and refers to this issue, before closing this issue.

**Comment by pixelzoom** [comment]
@jonathanolson identified the general problem in https://github.com/phetsims/scenery-phet/issues/488#issuecomment-483738588:
> It looks like Shape.transformed is not behaving correctly for a certain class of transforms (depending on the exact scale/translation) ....

Recommended to have him create an issue for that, including any additional information that he knows about.  In face-to-face discussion, I recall him providing more details that I don't see documented in in this issue.

**Comment by jonathanolson** [comment]
Added details, image and a link in https://github.com/phetsims/kite/issues/77. Closing here.
